miniapps/autodiff/par_example.cpp

```diff
@@ -0,0 +1,521 @@
+   nr=new ParNLSolverPLaplacian(MPI_COMM_WORLD,*pmesh, fespace, 2.0, &load);
+   nr->SetIntegrator(integrator);
+   nr->SetMaxNRIter(newton_iter);
+   nr->SetNRATol(newton_abs_tol);
+   nr->SetNRRTol(newton_rel_tol);
+   nr->SetPrintLevel(print_level);
+   timer->Clear();
+   timer->Start();
+   nr->Solve(*sv);
+   timer->Stop();
+   if (myrank==0)
+   {
+      std::cout << "[pp=2] The solution time is: " << timer->RealTime()
+                << std::endl;
+   }
+   // Compute the energy
+   double energy = nr->GetEnergy(*sv);
+   if (myrank==0)
+   {
+      std::cout << "[pp=2] The total energy of the system is E=" << energy
+                << std::endl;
+   }
+   delete nr;
+   x.SetFromTrueDofs(*sv);
+   dacol->SetTime(2.0);
+   dacol->SetCycle(2);
+   dacol->Save();
+
+
+   // 12. Continue with powers higher than 2
+   for (int i = 3; i < pp; i++)
+   {
+      nr=new ParNLSolverPLaplacian(MPI_COMM_WORLD,*pmesh, fespace, (double)i, &load);
+      nr->SetIntegrator(integrator);
+      nr->SetMaxNRIter(newton_iter);
+      nr->SetNRATol(newton_abs_tol);
+      nr->SetNRRTol(newton_rel_tol);
+      nr->SetPrintLevel(print_level);
+      timer->Clear();
+      timer->Start();
+      nr->Solve(*sv);
+      timer->Stop();
+      if (myrank==0)
+      {
+         std::cout << "[pp="<<i<<"] The solution time is: " << timer->RealTime()
+                   << std::endl;
+      }
+      // Compute the energy
+      double energy = nr->GetEnergy(*sv);
+      if (myrank==0)
+      {
+         std::cout << "[pp="<<i<<"] The total energy of the system is E=" << energy
+                   << std::endl;
+      }
+      delete nr;
+      x.SetFromTrueDofs(*sv);
+      dacol->SetTime((double)i);
+      dacol->SetCycle(i);
+      dacol->Save();
+   }
+
+   // 13. Continue with the final power
+   if (std::abs(pp - 2.0) > std::numeric_limits<double>::epsilon())
+   {
+      nr=new ParNLSolverPLaplacian(MPI_COMM_WORLD,*pmesh, fespace, pp, &load);
+      nr->SetIntegrator(integrator);
+      nr->SetMaxNRIter(newton_iter);
+      nr->SetNRATol(newton_abs_tol);
+      nr->SetNRRTol(newton_rel_tol);
+      nr->SetPrintLevel(print_level);
+      timer->Clear();
+      timer->Start();
+      nr->Solve(*sv);
+      timer->Stop();
+      if (myrank==0)
+      {
+         std::cout << "[pp="<<pp<<"] The solution time is: " << timer->RealTime()
+                   << std::endl;
+      }
+      // Compute the energy
+      double energy = nr->GetEnergy(*sv);
+      if (myrank==0)
+      {
+         std::cout << "[pp="<<pp<<"] The total energy of the system is E=" << energy
+                   << std::endl;
+      }
+      delete nr;
+      x.SetFromTrueDofs(*sv);
+      dacol->SetTime(pp);
+      if (pp < 2.0)
+      {
+         dacol->SetCycle(std::floor(pp));
+      }
+      else
+      {
+         dacol->SetCycle(std::ceil(pp));
+      }
+      dacol->Save();
+   }
+
+   // 14. Free the used memory
+   delete dacol;
```

Comment:
std::unique_ptr instead?

Reply:
I would prefer to keep the old fashion way in the examples in order to keep the focus on the AD instead of adding "new" techniques to the code. I will use smart pointers in the next example.